### PR TITLE
[RFC] do not require pid for numeric article ids

### PR DESCRIPTION
**Pull request by fritzmg**
During refactoring of one of my extensions I noticed the following: currently is impossible to do something as simple as:
```php
<?= \Contao\Controller::getArticle(2) ?>
```
This code only works on the page to which the article with ID _2_ belongs to.

Why? `Controller::getArticle` uses

https://github.com/contao/contao/blob/ab11d8663101236b68f72bcb96de9146f88658eb/core-bundle/src/Resources/contao/library/Contao/Controller.php#L359-L379

And `ArticleModel::findByIdOrAliasAndPid` will _always_ add the `pid` to the query, when it is present:

https://github.com/contao/contao/blob/ab11d8663101236b68f72bcb96de9146f88658eb/core-bundle/src/Resources/contao/models/ArticleModel.php#L133-L146

Since `Controller::getArticle` only passes `null` for `$intPid` when it is used as an insert tag and otherwise always passes the ID of the _current_ page, it is impossible to retreive the full markup of an article (including its wrapper) on any other page via `Controller::getArticle`.

This seems like an unecessary limitation. If you have the numeric ID of an article, the parent ID does not matter at all. In my extension I work around that problem by doing something like this:
```php
// Override global page object
global $objPage;
$currentPage = $objPage;
$article = $articleModel->findById($articleId);
$objPage = $pageModel->findById($article->pid);

// Get the fully rendered article
$renderedArticle = $controller->getArticle($articleId, false, false, $column);

// Reset global page object
$objPage = $currentPage;
```
This PR would fix the limitation directly within `ArticleModel::findByIdOrAliasAndPid`. However we could also do
```diff
-   $objRow = \ArticleModel::findByIdOrAliasAndPid($varId, (!$blnIsInsertTag ? $objPage->id : null));
+   $objRow = \ArticleModel::findByIdOrAliasAndPid($varId, (!$blnIsInsertTag && !\is_numeric($varId) ? $objPage->id : null));
```
within `Controller::getArticle`. Though I think the former makes more sense since, as I mentioned before, the pid does not matter, if you have the numeric ID of an article.

Besides, I interpret the meaning behind the function name `findByIdOrAliasAndPid` as:

find by `id || alias && pid`

i.e., find by either an ID … _or_ an alias __and__ a PID.

### CODE REVIEW

**Comment by leofeyer** [comment]
The logic is `(id || alias) && pid` though and not `id || (alias && pid)`, therefore we should do something like

```php
if (is_numeric($varId)) {
    $row = ArticleModel::findByPk($varId);
} else {
    $row = ArticleModel::findByIdOrAliasAndPid($varId, (!$blnIsInsertTag ? $objPage->id : null));
}
```

**Comment by fritzmg** [comment]
> The logic is `(id || alias) && pid`

I know, that's the current state. But you do not need a `pid` for an `id` - the `id` is unique across all `tl_article` entries. You only need the `pid` when using the `alias`, since there can be multiple `tl_article` entries for the same `alias` (I think? In multidomain installations?). Or what would be the purpose of querying the database for an additional parent ID, if you already have the ID?

**Comment by leofeyer** [comment]
> purpose of querying the database for an additional parent ID, if you already have the ID?

There is none, because the ID is unique. But if we decide to change this, we would have to change it everywhere and not just in one single method. And there is no guarantee that it does not imply a BC break.

**Comment by fritzmg** [comment]
True, it would imply a BC break... someone could rely on there _not_ being a result, if the pid does not fit the alias or id 😁

**Comment by fritzmg** [comment]
Two methods would be affected. `ArticleModel::findByIdOrAliasAndPid` and `ArticleModel::findPublishedByIdOrAliasAndPid`. No other Contao Core model has an "id or alias and pid" method.

**Comment by leofeyer** [comment]
I would not risk a BC break. @contao/developers WDYT?

**Comment by ausi** [comment]
Could we add a new method to the model, use that one in `Controller::getArticle()` and deprecate the old one?

**Comment by fritzmg** [comment]
The BC break would still apply to `Controller::getArticle()`.

**Comment by aschempp** [comment]
The best backwards compatible option I can think of is something like allow `null` for `$blnIsInsertTag` and add a the model query to `($blnIsInsertTag === false ? $objPage->id : null)`

**Comment by aschempp** [comment]
@fritzmg will you update the PR according to my comment?

**Comment by fritzmg** [comment]
I will, yes :)

**Comment by fritzmg** [comment]
I have pushed the changes. Doing something like
```php
<?= \Contao\Controller::getArticle(2, false, null) ?>
```
will give you the full HTML markup of the article with ID 2, independent from the current `$objPage` object.

**Comment by leofeyer** [comment]
> will give you the full HTML markup of the article with ID 2

According to your phpDoc comment, it will give you the article with "no HTML markup for the container". What is correct?

And what if I want an article independent from the current page but **with** the full HTML markup?

**Comment by fritzmg** [comment]
> According to your phpDoc comment, it will give you the article with "no HTML markup for the container". What is correct?

"no HTML markup for the container" only applies when the third parameter `$blnIsInsertTag` is `true`. In all other cases the HTML markup is not omitted.

**Comment by leofeyer** [comment]
I see. However, that is really ugly. Why did you not add a fourth argument?

**Comment by fritzmg** [comment]
You mean a fifth, like `$blnIgnorePageRelation` and
```php
($blnIsInsertTag === false && !$blnIgnorePageRelation ? $objPage->id : null)
```
which would lead to
```php
<?= \Contao\Controller::getArticle(2, false, false, 'main', true) ?>
```
? Sure, I don't really care. Imho this is just as ugly :grin:. Personally i'd still rather change the `ArticleModel` functions, because getting the fulll markup of an article should ideally be as easy as
```php
<?= \Contao\Controller::getArticle(2) ?>
```
But we cannot do that without breaking BC.

**Comment by leofeyer** [comment]
Let's discuss the implementation in the next Mumble call.

**Comment by leofeyer** [comment]
As discussed in Mumble on July 4th, we want to add a fifth argument to the existing method and a new method `getArticleById()` which calls `getArticle()` with the correct arguments.

**Comment by fritzmg** [comment]
Sounds good, I'll adjust the PR accordingly.

File: core-bundle/src/Resources/contao/library/Contao/Controller.php [hunk 1]
```diff
@@ -422,6 +423,21 @@ public static function getArticle($varId, $blnMultiMode=false, $blnIsInsertTag=f
...
+	 * Generate an article and return it as string
+	 *
+	 * @param mixed   $intId          The article ID
+	 * @param boolean $blnMultiMode   If true, only teasers will be shown
+	 * @param boolean $blnIsInsertTag If true, there will be no page relation and no HTML markup for the container
+	 * @param string  $strColumn      The name of the column
+	 *
+	 * @return string|boolean The article HTML markup or false
+	 */
+	public static function getArticleById($intId, $blnMultiMode=false, $blnIsInsertTag=false, $strColumn='main')
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: I think only the `$intId` parameter should be used here, without the other three parameters.

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: But then you will not be able to control whether you want the container markup. The `$strColumn` variable on the other hand seems to be a left over from the past. It is passed to the `ModuleArticle` instance which in turn passes it to the template. But the current default `mod_article` template does not use this variable.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: Then why did you add it to the new method?

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: Ah, it is used here:

https://github.com/contao/contao/blob/26b355525b045fedf129f55e48253d75c8cedb91/core-bundle/src/Resources/contao/modules/ModuleArticle.php#L193

Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: > Then why did you add it to the new method?

I only looked at it when @ausi questioned the addition of all parameters ;)



> Ah, it is used here:

Hm, same story there though. It is passed to the template, but not used by default anywhere. But some people might have custom content element template and check on the column.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Controller.php:436]: I think you have to search for "inColumn" to find its usage, e.g. here:

https://github.com/contao/contao/blob/b48f650dd49a58daae5883ded66aac311ccffd75/core-bundle/src/Resources/contao/elements/ContentTeaser.php#L75-L78

**Review by leofeyer — CHANGES_REQUESTED** [review]
Since it seems that we cannot get rid of the method arguments, I vote for removing the new `getArticleById()` method again. A fifth argument is just as ugly IMHO. 😄

**Comment by fritzmg** [comment]
I was able to solve my original problem in another way - since the `getArticle` method also accepts a complete `ArticleModel` object which completely circumvents the issue (see https://github.com/fritzmg/contao-inherit-article/commit/98048918526477681ab1a044509301279bc56bae).

So as a workaround you can also do the following:
```php
<?= \Contao\Controller::getArticle(\Contao\ArticleModel::findById(2)) ?>
```
Which I find at least better than ugly additional parameters ;). Thus currently I would vote for: leave everything as it is & change everything in Contao 5, if still applicable …

**Comment by leofeyer** [comment]
That seems to be the best solution indeed. 😄

**Pull request closed without merging**